src/reynir_correct/pattern.py

```diff
@@ -285,40 +296,67 @@ def wrong_preposition_vitni_af(self, match: SimpleTree) -> None:
             )
         )
 
-    def wrong_preposition_heillaður_að(self, match: SimpleTree) -> None:
+    def wrong_preposition_grin_af(self, match: SimpleTree) -> None:
         """ Handle a match of a suspect preposition pattern """
+        # Find the offending verbal and nominal phrases
+        vp = match.first_match("VP > { 'gera' }")
+        np = match.first_match("NP > { 'grín' }")
+        if np is None:
+            np = match.first_match("NP >> { 'grín' }")
+        # Find the attached prepositional phrase
+        pp = match.first_match('P > { "af" }')
+        if pp is None:
+            pp = match.first_match('ADVP > { "af" }')
+        assert np is not None
+        assert pp is not None
         # Calculate the start and end token indices, spanning both phrases
-        start, end = match.span
-        text = "'heillaður að' á sennilega að vera 'heillaður af'"
+        if vp is None:
+            start, end = min(np.span[0], pp.span[0]), max(np.span[1], pp.span[1])
+        else:
+            start, end = min(vp.span[0], np.span[0], pp.span[0]), max(vp.span[1], np.span[1], pp.span[1])
+        text = "'gera grín af' á sennilega að vera 'gera grín að'"
         detail = (
-            "Í samhenginu 'heillaður af e-u' er notuð " "forsetningin 'af', ekki 'að'."
+            "Í samhenginu 'gera grín að e-u' er notuð "
+            "forsetningin 'að', ekki 'af'."
         )
-        if match.tidy_text.count(" að ") == 1:
-            # Only one way to substitute að -> af: do it
-            suggest = match.tidy_text.replace(" að ", " af ")
+        if match.tidy_text.count(" af ") == 1:
+            # Only one way to substitute af -> að: do it
+            suggest = match.tidy_text.replace(" af ", " að ")
         else:
             # !!! TODO: More intelligent substitution to create a suggestion
             suggest = ""
         self._ann.append(
             Annotation(
                 start=start,
                 end=end,
-                code="P_WRONG_PREP_AÐ",
+                code="P_WRONG_PREP_AF",
                 text=text,
                 detail=detail,
-                original="að",
+                original="af",
```

Comment:
`original=match.tidy_text` ?  At least if `suggest` is not an empty string.

Reply:
I am kind of wondering about the purpose of the "original" parameter, since we always have the original span available through the token list and the start and end indices. Don't bother fixing this for now, we'll look at it in a broader context.